staker/block_validator.go

```diff
@@ -490,527 +426,436 @@ func (v *BlockValidator) SetCurrentWasmModuleRoot(hash common.Hash) error {
-			}
-		}
-		nextRecord++
 	}
+	nextPrepared := v.validated() + arbutil.MessageIndex(v.config().PrerecordedBlocks)
+	created := v.created()
+	if nextPrepared > created {
+		nextPrepared = created
+	}
+	if v.prepared >= nextPrepared {
+		return nil
+	}
+	nextPromise := containers.NewPromise[arbutil.MessageIndex](nil)
+	v.LaunchThread(func(ctx context.Context) {
+		err := v.recorder.PrepareForRecord(ctx, v.prepared, nextPrepared-1)
+		if err != nil {
+			nextPromise.ProduceError(err)
+		} else {
+			nextPromise.Produce(nextPrepared)
+			nonBlockingTriger(v.sendRecordChan)
+		}
+	})
+	v.nextRecordPrepared = &nextPromise
+	return nil
 }
 
-func (v *BlockValidator) writeLastValidatedToDb(blockNumber uint64, blockHash common.Hash, endPos GlobalStatePosition) error {
-	info := LastBlockValidatedDbInfo{
-		BlockNumber:   blockNumber,
-		BlockHash:     blockHash,
-		AfterPosition: endPos,
+func (v *BlockValidator) sendNextRecordRequest(ctx context.Context) (bool, error) {
+	v.reorgMutex.RLock()
+	defer v.reorgMutex.RUnlock()
+	err := v.sendNextRecordPrepare()
+	if err != nil {
+		return false, err
+	}
+	pos := v.recordSent()
+	if pos >= v.prepared {
+		log.Trace("next record request: nothing to send", "pos", pos)
+		return false, nil
+	}
+	validationStatus, found := v.validations.Load(pos)
+	if !found {
+		return false, fmt.Errorf("not found entry for pos %d", pos)
 	}
-	encodedInfo, err := rlp.EncodeToBytes(info)
+	currentStatus := validationStatus.getStatus()
+	if currentStatus != Created {
+		return false, fmt.Errorf("bad status trying to send recordings for pos %d status: %v", pos, currentStatus)
+	}
+	err = v.sendRecord(validationStatus)
 	if err != nil {
-		return err
+		return false, err
 	}
-	err = v.db.Put(lastBlockValidatedInfoKey, encodedInfo)
+	atomicStorePos(&v.recordSentA, pos+1)
+	log.Trace("next record request: sent", "pos", pos)
+	return true, nil
+}
+
+func (v *BlockValidator) iterativeValidationEntryRecorder(ctx context.Context, ignored struct{}) time.Duration {
+	moreWork, err := v.sendNextRecordRequest(ctx)
 	if err != nil {
-		return err
+		log.Error("error trying to record for validation node", "err", err)
 	}
-	return nil
+	if moreWork {
+		return 0
+	}
+	return v.config().ValidationPoll
 }
 
-func (v *BlockValidator) progressValidated() {
-	v.reorgMutex.Lock()
-	defer v.reorgMutex.Unlock()
-	for atomic.LoadInt32(&v.reorgsPending) == 0 {
-		// Reads from blocksValidated can be non-atomic as all writes hold reorgMutex
-		checkingBlock := v.lastBlockValidated + 1
-		entry, found := v.validations.Load(checkingBlock)
-		if !found {
-			return
-		}
-		validationStatus, ok := entry.(*validationStatus)
-		if !ok || (validationStatus == nil) {
-			log.Error("bad entry trying to advance validated counter")
-			return
+func (v *BlockValidator) maybePrintNewlyValid() {
+	if time.Since(v.validInfoPrintTime) > time.Second {
+		log.Info("result validated", "count", v.validated(), "blockHash", v.lastValidGS.BlockHash)
```

Comment:
One thing I don't love about this is that the last "result validated" log line might be slightly behind the actual last validated block, because at the time the last block was validated there had already been another log under a second ago. That would make debugging tricky, because it might seem like the block validator got stuck and never validated the most recent block when it actually did. I think we should try to invoke this function regularly (and keep track of what we last logged so we don't log the same thing twice).